### DISCUSSION TITLE
Scaling issue? (Solved)

### DISCUSSION THREAD

**Discussion by 98mux** [pocketbase/pocketbase#5574] [Q&A] [answered]
Edit: My scaling issues were solved - see my last comment

Lately I've been getting 200-500 users a day for my app and I'm noticing scaling issues, especially queries which takes longer than 1 second. Most of my queries are simple single row get and post queries.

I have about 30 tables with several of them having over 100k+ rows and 10 000 users in total

I understand fully that pocketbase isn't built for scale and I'm preparing to migrate over to something else. But there are some things here that doesn't make sense to me - the most important one is the low VCPU usage I have on railway.app. I'm averaging just 0.1 VCPU - so maybe there is some kind of Disk I/O bottleneck here?

<img width="811" alt="Screenshot 2024-09-26 at 23 53 29" src="https://github.com/user-attachments/assets/e035a146-b84a-416e-97e2-27368e97c65e">

According to Railway they are using ZFS on fast NVME disks for data storage - so surely it should be fast? https://blog.railway.app/p/launch-01-next-gen-databases

My current load is about 13 000 requests per hour, or about 4 requests per second

<img width="1355" alt="Screenshot 2024-09-27 at 00 02 01" src="https://github.com/user-attachments/assets/d250f317-3697-49ad-b97b-22582cabef6d">

The same hour I had 13 000 requests I had 1 400 requests that took longer than 1000 ms, over 10%

<img width="1355" alt="Screenshot 2024-09-27 at 00 03 22" src="https://github.com/user-attachments/assets/dec75054-9258-46fb-83d7-ced57e4a139b">

Here are some requests from the peak hour that was really struggeling (data.execTime > 1000)
There are no obvious signs - just normal requests that are mostly get/post/patch of single rows. I've already optimized for indexes to improve performance

<img width="1319" alt="Screenshot 2024-09-27 at 00 07 31" src="https://github.com/user-attachments/assets/b0f13b20-6fe1-4c37-b1da-7af6b2300d34">

The pocketbase version is 0.22.19
A Pocketbase backup is around 400MB

My question is - how do i go about debugging this to figure out what the bottleneck is?

**Comment by ganigeorgiev** [comment]
I promised my self that I'll ignore this type of posts but I hate when people complains about performance or scaling without providing reproducible or anything concrete to discuss.

The amount of traffic you are experiencing doesn't seem that much. I'm using PocketBase in [Presentator](https://github.com/presentator/presentator) with heavier usage both in terms of number of users and executed requests (_at the moment the peak total number of requests per hour is ~24k; see also https://github.com/pocketbase/pocketbase/discussions/4254_) and from a quick search in the logs from the past week I don't have a write or read response that takes more than 1500ms (the majority of the read requests are between 0~400ms and the writes are between 0-20ms; the only occasionally slightly slower GET requests are file download requests because of large image thumbs generation on the fly).

In any case, the number of users or requests doesn't really tells much.

**Having a write request (PATCH/POST) that takes more than 1s is definitely not normal** but what could be the cause is difficult to guess (_excluding users create requests due to the bcrypt hash generation_).

You've mentioned that "I've already optimized for indexes" but how are you sure that your indexes actually helped and not make things worse (_especially if the defined indexes are composite; see https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045_)?

I don't think I can help you without details about your collections structure and the data you are storing, but I'd recommend to:

- ensure that you don't have slow blocking tasks as part of a custom hook that interferes with the default operation (e.g. sending an email or request to external service after each record create)

- check your Create/Update API rules (_excessive usage of `@collection.*` or other fields that perform joins over large datasets can hurt the performance_)

- avoid slow tasks when inside `RunInTransaction` calls (if you have any)

- if you are storing large JSON data that you are not accessing/updating frequently it may be beneficial to split it in separate collection

- try to reproduce the issue locally, start the PocketBase executable with `--dev` and inspect the queries printed in the terminal

Your problem doesn't seem caused by scaling and it will not magically disappear by adding another server or switching to something else.

I'm not ruling out a PocketBase issue or inefficient query, but without anything concrete there is nothing really I (and maybe others) that can assist you here.

**Reply by ganigeorgiev** [reply]
I've also noticed from the last logs screenshot that most of the slow requests seem to have happened around the same time so if the slow execution is not something that you experience regularly then it is possible to be caused by some temp availability issue with the VM resources (to rule this out try to reproduce it locally against a backup).

**Reply by dschissler** [reply]
@98mux I don't know why you would immediately assume that PB is the issue when you only have a problem in one small window of a few hours on a particular day.  You might as well disregard all of the other data and just drill into that small time frame.

Maybe you were using all of your cores for a background system process, or the provider was cheating you for a bit to do some maintenance, or they oversold their capacity and then had to remedy it.

Otherwise it seems like PB is showing almost zero signs of scaling issues.

[edit] Maybe another user (or yourself) was doing some sort of data backup on the shared NVME, either within the VM or at the provider level.

**Reply by 98mux** [reply]
@ganigeorgiev I want to begin by letting you know that pocketbase is my favorite piece of technology ever, it is an absolute joy to use. The last thing I want to be is an annoyance to you, I only seek for help and advice, which you have given me! Thank you so much for taking your time to reply, I really appreciate it.

1. On indexes: I’ve mostly enabled indexes to solve queries that take too long. The table I have the most indexes on is called heroes and it has the indexes “day”, “lastCheckin”, “day, lastCheckin”. I’ll remove “day, lastCheckin” and see if that helps
2. On custom hook: I have no extensions on the pocketbase instance. No custom js/go code
3. On API rules: Most of mye API rules are simply “@request.auth.id = userId” and “@request.auth.id = userId &&
(@request.data.userId = null || @request.data.userId = userId)” for update
4. On RunInTransaction: I don’t think I’ve used that.
5. On JSON: I do use JSON here and there, and I have discovered one JSON array that can grow very large (I’ve seen some with 15KB+). I’ll look more into this

What I do have is a vercel nodejs application that is logged in as an admin. ("Include requests by 
Admin" is turned on in the logs).

Heroes and users tables are often some of the slow queries. Some info about them is that users has a whopping 40 columns and heroes has 23 columns + a potentially large JSON field

Based on your feedback there is probably some chance that railway is the culprit. So in conclusion I’m going to migrate from railway to digital ocean and see if that solves the issue. Again, thank you so much for your help!

I will also try debugging locally as you said

**Reply by 98mux** [reply]
@dschissler Good to know that the PB isn't showing any signs of scaling issues. I have not "immediately assume that PB is the issue".
I have drilled into that small timeframe, you can see it in the last screenshot.
If you look away from the spike - there are still consistently around 40 requests an hour that are above 1000 ms, and they look simular to the spike.

Note there was also multiple more spikes 7+ days ago, which are not shown in the logs because that is over the timeframe of the logs

**Reply by ganigeorgiev** [reply, marked as answer]
> Some info about them is that users has a whopping 40 columns and heroes has 23 columns + a potentially large JSON field

Large number of fields or data in a single record could have impact on the performance simply due to copying (both implicit and explicit). If you can consistently reproduce the slow execution times I recommend to split the fields that are not frequently read/filtered/modified to a separate collection that can have a single `relation` referencing the main one.

> Based on you’re feedback there is probably some chance that railway is the culprit.

I'm not saying this but you shouldn't expect 100% consistent performance from a VM with shared resources.

Also note that as far as I understand Railway relies on volumes for persistence which although internally they might be backed by fast NVME disks it is still a virtual/network mounted disk and it will be almost always slower compared to a direct disk access like most regular VPS comes with. For example, based on https://pcr.cloud-mercato.com/providers/hetzner/flavors/cpx11/performance/storage-bandwidth, for Hetzner the default VPS local disk performs almost 10x more read operations compared to a mounted volume (_writes are more or less identical_). For other providers I imagine it would be similar.

In all cases, before changing providers I strongly recommend to try to reproduce the issue locally by running with `--dev` against a production backup and try to identify the slow queries.

**Reply by 98mux** [reply]
@ganigeorgiev Seems like I’ve solved it!!!

I did what you told me to, to test it locally, I also tested it on a fresh instance with the production backup on digital ocean and railway

Here were some tests I ran, I did 4000 requests (1000 getOne users, 1000 getOne heroes, 1000 update users, 1000 update heroes).

Digital ocean had only 1 VCPU, and local machine is a macbook M1. Railway and Digital ocean also have additional network latency as the calls were made from my laptop

**20 concurrent, 4000 requests**

Digital Ocean:
30 seconds total
0 requests above 1 second, 125 requests/s

Local machine:
8 seconds total
0 requests above 1 second, 450 requests/s

Railway:
15 seconds total
0 requests above 1 second, 300 requests/s

**50 concurrent, 4000 requests**

Digital Ocean:
36 seconds total
378 requests above 1s, 100 requests/s

Local machine:
10 seconds total
49 requests above 1s, 400 requests/s,

Railway:
20 seconds total
38 requests above 1s, 200 requests/s

Railway performed really good in the benchmarks which made me question why on earth my production project struggled so bad

The solution:

My railway project had 5 GB diskspace and pocketbase was using 2.5 GB. When i tried to upgrade the storage to 50 GB then I got an error from Railway saying that it wasn’t able to upgrade the storage - which suggests that the volume is faulty of some kind. When i migrated over to a brand new project in railway with 50GB the issue was fixed and pocketbase is now using 3GB in disk space

I believe the issue I experienced was one of these four:

1. My project’s volume had a faulty disk from Railway
2. Railway might have given me a weaker volume to start with for the free/hobby tier compared to what i would get if i started the project in pro-tier
3. The SSD was shared by too many projects and some project was using up all of its I/O capacity
4. Pocketbase or sqlite might have some kind of configuration that limits it from using more than 50% of diskspace? Causing it to perform bad in low-diskpace situations as it tries to go below the limit

I’ve had no issues the past 3 hours, I hope I’m not calling this too early 😭 but it seems solid, time will tell

I’m so happy now!!! I can continue to use Pocketbase!!! 😍😍😍😍😍😍 Thank you so much for helping!

**Reply by dschissler** [reply]
Are the requests that were taking more than 1s from queries across the board, or is it only specific queries which sometimes become slow?  You could document this while it's still fresh in your memory.  Whatever disk problem was affecting you could become relevant later were you to grow a lot, or need to switch hosts.

**Reply by 98mux** [reply]
@dschissler Across the board - like hiccups. It is as random as it can be, queries that usually take 0.5ms suddenly takes 20 000 ms

Still looking good, but there has been a super small hiccup or two after the upgrade - but about 99.5% less than before. The flatline is after i migrated from railway to railway. (The logs are "data.execTime > 1000 && data.url !~ "avatar" && data.url !~ "auth")
<img width="935" alt="Screenshot 2024-09-28 at 13 48 09" src="https://github.com/user-attachments/assets/96741b2d-bbe9-4d88-832d-48aca28ef050">

**Reply by dschissler** [reply]
@98mux Maybe you could run a logging program at the OS level for memory, disk and cpu utilization.  Then look at hotspots in your pocketbase logs and then manually compare it to this period in your OS logs.  If you don't notice anything then maybe it's all occurring below at the virtualization layer (backups or other VMs).

**Reply by 98mux** [reply]
@dschissler That's a good idea! I've never done something like that before - do you recommend any program like that? and that maybe has docker image to go along that?

**Reply by whmeitzler** [reply]
> @dschissler That's a good idea! I've never done something like that before - do you recommend any program like that? and that maybe has docker image to go along that?

Glad to see Pocketbase is working for you. If you want a simple drop-in tool for server monitoring, I haven't found anything better (or easier to use) than [Netdata](https://github.com/netdata/netdata). It's free, C, and on apt-get.

**Reply by 98mux** [reply]
@whmeitzler Awesome! Thanks for the suggestion! I'll look into it :D
